Add image verifier transfer service plugin system based on a binary directory

ethan-lowman-dd opened the pull request:
Implements https://github.com/containerd/containerd/issues/6691

See the added docs for configuration details. To test, run a command like `ctr image pull --local=false index.docker.io/library/alpine@sha256:c75ac27b49326926b803b9ed43bf088bc220d22556de1bc5f72d742c91398f69`.

The simplest verifier binaries for testing can be something like this:
```
#!/usr/bin/env bash

echo Approved
```

or 

```
#!/usr/bin/env bash

echo Rejected
exit 1
```

Real implementations would need to parse the `-name` and `-digest` flags.

ktock commented on docs/image-verification.md:
````diff
@@ -0,0 +1,23 @@
+```yaml
+[plugins]
+  [plugins."io.containerd.image-verifier.v1.bindir"]
+    enabled = true
````
Are there any use cases where the plugin is registered but `enabled = false`?

ktock commented on docs/image-verification.md:
```diff
@@ -0,0 +1,23 @@
+
+## Image Pull Judgement
+
+Return an exit code of 0 to allow the image to be pulled and any other exit code to reject the image.
```
What is the expected verification the plugin performs based on `name` and `digest`?

ethan-lowman-dd replied:
They can do essentially anything, such as:
1. compare registry against a trusted list of registries
2. block the `latest` tag
3. block pulling images without digest references.
4. check image signatures (this would probably be the most popular use case -- see https://github.com/containerd/containerd/issues/6691 for details)

ktock reviewed (commented):
Is this CRI-specific feature (i.e. should be implemented under `pkg/cri`) or does this expand containerd's scope?

ethan-lowman-dd commented:
> Is this CRI-specific feature

@ktock I had previously prototyped an implementation of this feature in the CRI (here: https://github.com/containerd/containerd/pull/6994) but the feedback from maintainers was that it should live somewhere closer to the core so all containerd clients can benefit, not just the CRI. In particular, the Transfer service was named as the ideal integration spot.

dmcgowan commented on pkg/imageverifier/image_verifier.go:
```diff
@@ -0,0 +1,29 @@
+import "context"
+
+type ImageVerifier interface {
+	Enabled(ctx context.Context) bool
```
How is `Enabled` intended to be used here. There is already a built-in way to disable plugins in containerd and plugins can also disable themselves during Init.

ethan-lowman-dd replied:
Similar thread here: https://github.com/containerd/containerd/pull/8493#discussion_r1188281709

These logs might be noisy if the user hasn't placed any verifier binaries, but would be useful for audit purposes:
```
INFO[2023-05-10T22:16:29.097720765Z] Verifying image pull                          digest="sha256:02bb6f428431fbc2809c5d1b41eab5a68350194fb508869a33cb1af4444c9b11" name="index.docker.io/library/alpine:latest#@sha256:c75ac27b49326926b803b9ed43bf088bc220d22556de1bc5f72d742c91398f69" verifier=bindir
INFO[2023-05-10T22:16:29.153495401Z] Image verifier allowed pull                   digest="sha256:02bb6f428431fbc2809c5d1b41eab5a68350194fb508869a33cb1af4444c9b11" name="index.docker.io/library/alpine:latest#@sha256:c75ac27b49326926b803b9ed43bf088bc220d22556de1bc5f72d742c91398f69" ok=true reason="accepted" verifier=bindir
```
The idea of was to avoid logging noise for the default image verifier configuration (i.e. no binaries in the verifier directory) by requiring opt-in via `Enabled` rather than opt-out via the `disabled_plugins` config field. If these logs aren't an issue, I can just remove `Enabled` from the interface. If the logs are an issue, we could remove `Enabled` and push the logs into the plugin implementation. Alternatively, users could just opt-out using the regular `disabled_plugins` mechanism if they don't like the logs.

samuelkarp replied:
Can we move these logs to debug level?  I'd expect a denied pull to show up in logs, but not necessarily an allowed pull.

dmcgowan commented on pkg/imageverifier/image_verifier.go:
```diff
@@ -0,0 +1,29 @@
+
+type ImageVerifier interface {
+	Enabled(ctx context.Context) bool
+	VerifyImage(ctx context.Context, name string, digest string) Judgement
```
I like the simplicity but we try to use the full descriptor object over a raw digests in most interfaces. It avoids the inevitable case when media-type/size/annotations want to be used by implementations.

samuelkarp commented on docs/image-verification.md:
````diff
@@ -0,0 +1,23 @@
+    bin_dir = "/opt/containerd-image-verifier/bin"
+```
+
+Place one or more verifier binaries in `bin_dir` which conform to the following API:
````
We should be explicit over the order in which binaries are invoked (which might be "no order" and we execute in parallel).

samuelkarp replied:
From a CRI perspective: binary plugins like this will have their resources accounted in (and constrained by) containerd's cgroups rather than a pod's.  I don't think we need to solve this in the first iteration, but we may want to think about possibilities for charging resource usage to a pod.

samuelkarp commented on pkg/imageverifier/bindir/bindir.go:
```diff
@@ -0,0 +1,96 @@
+			"-digest", digest,
+		}
+
+		cmd := exec.CommandContext(ctx, bin, args...)
```
We should probably set a timeout on this command execution.  (That might be something worth making configurable too.)

samuelkarp commented:
```
   * 002cf10b "Prototype image verifier transfer service plugin system based on a binary directory" ... FAIL
    - PASS - commit does not have any whitespace errors
    - FAIL - does not have a valid DCO
    - PASS - commit subject is under 90 characters, but is still more than 72 chars
```

CI is failing since the commit is missing a sign-off.  See the [contribution guidelines](https://github.com/containerd/project/blob/main/CONTRIBUTING.md#sign-your-work) for instructions.

ethan-lowman-dd commented on docs/image-verification.md:
````diff
@@ -0,0 +1,23 @@
+
+```yaml
+[plugins]
+  [plugins."io.containerd.image-verifier.v1.bindir"]
````
As far as I'm can tell, the string `bindir` in this context corresponds with the ID of the registered plugin here:

```go
func init() {
	plugin.Register(&plugin.Registration{
		Type:   plugin.ImageVerifierPlugin,
		ID:     "bindir",
		Config: defaultConfig(),
		InitFn: func(ic *plugin.InitContext) (interface{}, error) {
			cfg := ic.Config.(*bindir.Config)
			return bindir.NewImageVerifier(cfg), nil
		},
	})
}
```

So I think this is not user configurable -- does that sound correct?

samuelkarp replied:
Yes, thanks.  I misunderstood what your intent was here.

samuelkarp commented on pkg/imageverifier/image_verifier.go:
```diff
@@ -0,0 +1,29 @@
+	VerifyImage(ctx context.Context, name string, digest string) Judgement
+}
+
+type Judgement struct {
+	OK     bool
+	Reason string
+}
```
Got it, the "verifier verify.sh rejected image" string comes as the `error`.

ethan-lowman-dd commented on pkg/imageverifier/image_verifier.go:
```diff
@@ -0,0 +1,32 @@
+)
+
+type ImageVerifier interface {
+	VerifyImage(ctx context.Context, name string, desc ocispec.Descriptor) Judgement
```
Looking for feedback here: I chose not to include an error in the return here so plugin implementations have to decide for themselves whether they want to fail open or fail closed on errors. Does that sound reasonable?

samuelkarp replied:
I have a weak opinion for including an `error`.  An `error` during execution of a verifier (for example with bindir, encountering a permission error attempting to run an executable) is different from the logic of a verification plugin issuing an explicit judgement, and an `error` allows that to be dealt with in the calling code.  An implementation of an `ImageVerifier` that chooses to fail-open might just not emit an error, or we could define an error type that is a fail-open error.

ethan-lowman-dd replied:
Added an error in 0ef20624faa5f60e1778e2be4e42632eeb81b9f7.

ethan-lowman-dd commented on pkg/imageverifier/bindir/bindir.go:
```diff
@@ -0,0 +1,189 @@
+	binPath := path.Join(v.config.BinDir, bin)
+	args := []string{
+		"-name", imageName,
+		"-digest", desc.Digest.String(),
```
Looking for feedback: should we include other descriptor fields? There is a lot of data in there, but I think most verifiers would just use the digest.

samuelkarp replied:
I think most of the fields would be useful, with the exception of `URLs` and `Data`.  `Size` is maybe another questionable one since that'd just be the size of the index or manifest itself.

* `MediaType` and `ArtifactType` - both help determine the content of the thing being pulled, and a verifier might want to limit that type of content
* `Annotations` - Could be used to allow or deny some images based on how they were built, such as if an image was generated from a CI system.  Combined with a signature over the manifest, an annotation could be attested data.
* `Platform` - Could be used as an easy way to restrict containerd's automatic platform fallback for Arm versions, restrict Linux images from being run on FreeBSD, etc.

ethan-lowman-dd replied:
I will add those. 

> Could be used as an easy way to restrict containerd's automatic platform fallback for Arm versions, restrict Linux images from being run on FreeBSD, etc.
 
Aside: I've personally seen a case where the automatic fallback from `amd64` to `i386` causes a huge performance degradation and it would be useful to be able to restrict that behavior via this mechanism.

ethan-lowman-dd replied:
What are your thoughts on passing the descriptor via stdin in JSON format? This makes it difficult to evolve the API, since stdin will be pinned to a specific meaning, but some of the fields are awkward to pass via flags, like a string map for annotations.

samuelkarp replied:
stdin as a JSON object makes sense to me, though we could wrap it in some other object as future-proofing for more data to be passed in the future.

@dmcgowan I'd be interested in your thoughts here.

dmcgowan replied:
stdin makes sense to me as well. Ideally set a `--stdin-media-type` or something though to type the input object. Creating an OCI index as the input could make sense here as well.

ethan-lowman-dd replied:
Passing a Descriptor over stdin is implemented in e87da1322a70e0039f1964c5669e0031bf60b0eb

samuelkarp commented on plugins/imageverifier/plugin.go:
```diff
@@ -0,0 +1,45 @@
+
+func defaultConfig() *bindir.Config {
+	return &bindir.Config{
+		BinDir:             "/opt/containerd-image-verifier/bin",
```
On Linux, maybe `/opt/containerd/image-verifier`?  While I don't think this would work easily with the [managed opt plugin](https://github.com/containerd/containerd/blob/main/docs/managed-opt.md), we do already have `/opt/containerd` and it'd be nice in the future to possibly use this mechanism to install verifiers.

For Windows: we should either specify a relevant Windows directory or drop Windows support for the verifier for now (and address in a follow-up?).

ethan-lowman-dd replied:
I need to do some research on Windows file layouts (I have ~no experience with Windows) but this sounds like just a few lines to add Windows support to this PR.

ethan-lowman-dd replied:
Done in 7cccbad9ec0ce2ec82050aa4a772a0e9cb3a297b

samuelkarp commented on docs/image-verification.md:
```diff
@@ -0,0 +1,39 @@
+- If `bin_dir` does not exist or contains no files, the image verifier does not block image pulls.
+- An image is pulled only if all verifiers that are called return an "ok" judgement (exit with status code 0). In other words, image pull judgements are combined with an `AND` operator.
+- If `max_verifiers < 0`, there is no imposed limit on the number of image verifiers called.
+- If `max_verifiers >= 0` (indicating there is a limit on the number of verifiers to call) and there are more than `max_verifier` entries in `bin_dir`, only `n = max_verifiers` of the verifiers will be called and the rest will be skipped. There is no guarantee which `n` of the verifiers are called.
```
Looking for opinions: should we lexicographically order the binaries and pick the first 10?

ethan-lowman-dd replied:
This is what the current implementation does, since `os.ReadDir` sorts by filename, but I wasn't sure if we wanted to commit to that implementation. If it's an observed consistent property of the implementation, perhaps it should be documented as such, however.

ethan-lowman-dd replied:
Addressed this in e146129ecb7bcdd601731dac0313db4ccee9035d

samuelkarp reviewed (commented):
I really like a lot of this.  As we get closer on a few of the open questions (and this moves from "WIP" to more ready to merge) we'll want to make sure we add some unit and integration tests.

akashsinghal commented:
Hi @ethan-lowman-dd :) Appreciate the work you're doing here on adding the verification capability. I actually built a very rudimentary prototype binary for Ratify for this and got it working with notation verifier configured. You can check it out [here](https://github.com/akashsinghal/ratify/pull/125). Thought I'd point you to it if you're curious. Ratify is a verification engine for supply chain artifacts and can be configured with many different verifiers (notation, cosign, spdx, sbom, & any custom verifier). It's been primarily used as a service working with an admission controller but the same principal could be applied here.

ethan-lowman-dd commented:
@samuelkarp @dmcgowan It seems like there's consensus on the design, so I went ahead and added tests, which takes this PR from WIP to something looking for a proper review.

tianon commented on plugins/imageverifier/path_windows.go:
```diff
@@ -0,0 +1,25 @@
+	"github.com/containerd/containerd/defaults"
+)
+
+var defaultPath = filepath.Join(defaults.DefaultRootDir, "opt")
```
This isn't quite specific enough, right?  This appears to be the same raw default as the `opt` plugin, and I think you probably meant to append `image-verifier/bin` to it just like the *nix path did:

```suggestion
var defaultPath = filepath.Join(defaults.DefaultRootDir, "opt", "image-verifier", "bin")
```

ethan-lowman-dd replied:
Fixed, thank you

ethan-lowman-dd commented:
@Vishal-Chdhry This plugin system does not assume that the verification involves signatures at all. There could be an implementation of a plugin that implements validation using a [policy.json](https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md) somewhere on disk, but I believe that would be outside the scope of containerd.

ethan-lowman-dd commented:
Rebased on main to try to get CI to pass (errors looked unrelated). Please let me know if there's any further feedback on this PR :)

dmcgowan commented:
@samuelkarp I think we can move forward and get this merged

samuelkarp commented:
/ok-to-test

RichardoC commented on pkg/imageverifier/bindir/bindir.go:
```diff
@@ -0,0 +1,243 @@
+	// Write the descriptor to stdin.
+	go func() {
+		// Descriptors are usually small enough to fit in a pipe buffer (which is
+		// often 64 KiB on Linux) so this write usually won't block on the child
```
I think there should be be a check before this that the descriptor is small enough to write, and exit this routine early if the write would fail

ethan-lowman-dd replied:
The write does not necessarily fail if the descriptor is larger than the pipe buffer -- it only fails if it's larger than the pipe buffer _and_ the child process doesn't read all of stdin, which is not something that can be predicted here -- that depends on the verifier's implementation. In the case of failure, a error is logged, but it doesn't affect the verification result. I added a test to demonstrate this in 3324f5c7fe1e5ba6d193ebe15508437d194e82c7

ethan-lowman-dd commented:
@dmcgowan Any idea why tests in https://github.com/containerd/containerd/tree/main/integration/client are failing? I tried a rebase several times.

dmcgowan commented:
@ethan-lowman-dd no, doesn't look related but looking into it before merging.

ethan-lowman-dd commented:
Ah I think you're right, it's failing (by chance, due to a lax string match) because a `sleep 1000` from a test verifier is still in the process tree. I tried disabling tests in https://github.com/containerd/containerd/pull/8493/commits/fde9525aa4891de8034088cd4c64e81975e9d45f to confirm, but I think I'll need to fix some kind of leak.

ethan-lowman-dd commented:
I think https://github.com/containerd/containerd/pull/8493/commits/1905eff9c279b0452b7f4ab6b27d6aa30e427421 should do the trick. I was killing the exec'ed process, but by default on Unix systems, that doesn't kill grand-child processes, leaving them orphaned if they don't exit on their own quickly. I added a unit test to check for orphaned processes similar to the one that was failing by chance in a different package.

Also, since I'm not super familiar with the Windows process model, I wanted to make sure the bindir tests ran on Windows too. I updated the testing strategy in `bindir_test.go` to use binaries built by the Go toolchain at test time instead of Bash scripts to support Windows. (Using the Go toolchain in Go tests may raise some eyebrows, but [it's something that the Go standard library does in a few places](https://cs.opensource.google/go/go/+/refs/tags/go1.20rc1:src/crypto/ecdh/ecdh_test.go;l=423-489).)

ethan-lowman-dd commented:
It might take me a bit of time to get this working on Windows (this is my first time building anything against Windows). I need to wrap my head around the differences in the Windows process model to avoid the two issues we have remaining affecting only Windows:
* when stdout reading is truncated, the child process is exiting with code 1
* child processes of verifiers are still leaking when verifiers are killed (having trouble finding a good way to kill a process and all its children on Windows)

ethan-lowman-dd commented:
I made an attempt to avoid orphan processes on Windows in this commit (not currently part of this PR):
https://github.com/ethan-lowman-dd/containerd/commit/0eaa502938f7cb9fefea50fedad10d63d5d27fb1

It's based on ideas I found here: https://devblogs.microsoft.com/oldnewthing/20131209-00/?p=2433

That commit doesn't current work -- `AssignProcessToJobObject` returns "Access Denied" and I'm not sure why. If there's someone who actually knows Windows who could give a hand, I'd really appreciate that :)

Or alternatively, we could let it be the case that orphan processes are a possibility on Windows. There doesn't seem to be good native Go support for avoiding that -- everything's through the Windows syscall package.

ethan-lowman-dd commented:
With the advice of some colleagues, I got things working on Windows in: https://github.com/containerd/containerd/pull/8493/commits/6e8376769d6eafe82c01a908b19a7ac095d9815a

ethan-lowman-dd commented:
/retest

ethan-lowman-dd commented:
The failing e2e tests looks to be this flake: https://github.com/kubernetes/kubernetes/issues/119600

But otherwise CI is green now. @dmcgowan This probably needs another look since I had to change a few things related to process management.

I will squash the commits, but I'll keep them for now so it's easy to see what changed since the first approvals.

dmcgowan commented:
/retest

dmcgowan commented:
@ethan-lowman-dd looks good, if the Windows stuff works, I'll believe the tests. Otherwise some Windows folks can comment on it. Can you squash some of those middle commits down and then I think it looks ready to go.

ethan-lowman-dd commented:
Squashed!
